pkg/admission/validate_backingstore.go

```diff
@@ -0,0 +1,321 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// const configuration values for the validation checks
+const (
+	MinimumVolumeSize       = 16
+	MaximumPvpoolNameLength = 43
+	MaximumVolumeCount      = 20
+)
+
+// BackingStoreValidator is the context of a backingstore validation
+type BackingStoreValidator struct {
+	BackingStore *nbv1.BackingStore
+	Logger       *logrus.Entry
+	arRequest    *admissionv1.AdmissionReview
+	arResponse   *admissionv1.AdmissionReview
+}
+
+// NewBackingStoreValidator initializes a BackingStoreValidator to be used for loading and validating a backingstore
+func NewBackingStoreValidator(arRequest admissionv1.AdmissionReview) *BackingStoreValidator {
+	bsv := &BackingStoreValidator{
+		arRequest: &arRequest,
+		Logger:    logrus.WithField("admission backingstore validation", arRequest.Request.Namespace),
+	}
+	return bsv
+}
+
+// ValidateBackingstore call appropriate validations based on the operation
+func (bsv *BackingStoreValidator) ValidateBackingstore() admissionv1.AdmissionReview {
+	bsv.arResponse = &admissionv1.AdmissionReview{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "AdmissionReview",
+			APIVersion: "admission.k8s.io/v1",
+		},
+		Response: &admissionv1.AdmissionResponse{
+			UID:     bsv.arRequest.Request.UID,
+			Allowed: true,
+			Result: &metav1.Status{
+				Message: "allowed",
+			},
+		},
+	}
+
+	switch bsv.arRequest.Request.Operation {
+	case admissionv1.Create:
+		bsv.ValidateCreate()
+	case admissionv1.Update:
+		bsv.ValidateUpdate()
+	case admissionv1.Delete:
+		bsv.ValidateDelete()
+	default:
+		bsv.Logger.Error("Failed to identify backindstore operation type")
+	}
+	return *bsv.arResponse
+}
+
+// RunValidation responsible of assinging the return values of a validation into the response appropriate fields
+func (bsv *BackingStoreValidator) RunValidation(isAllowed bool, message string) bool {
+	bsv.arResponse.Response.Allowed = isAllowed
+	bsv.arResponse.Response.Result.Message = message
+	return isAllowed
+}
+
+// DeserializeBS extract the backingstore object from the request
+func (bsv *BackingStoreValidator) DeserializeBS() *nbv1.BackingStore {
+	raw := bsv.arRequest.Request.Object.Raw
+	BS := nbv1.BackingStore{}
+	if err := json.Unmarshal(raw, &BS); err != nil {
+		bsv.Logger.Error("error deserializing backingstore")
+	}
+	return &BS
+}
+
+// DeserializeOldBS extract the old backingstore object from the request, exist only on UPDATE/DELETE requests
+func (bsv *BackingStoreValidator) DeserializeOldBS() *nbv1.BackingStore {
+	raw := bsv.arRequest.Request.OldObject.Raw
+	oldBS := nbv1.BackingStore{}
+	if err := json.Unmarshal(raw, &oldBS); err != nil {
+		bsv.Logger.Error("error deserializing old backingstore")
+	}
+	return &oldBS
+}
+
+// ValidateCreate runs all the validations tests for CREATE operations
+func (bsv *BackingStoreValidator) ValidateCreate() {
+	bsv.BackingStore = bsv.DeserializeBS()
+	if !bsv.RunValidation(bsv.ValidateBackingStoreType()) {
```

Comment:
I feel like this kind of writing is less understood to the reader than for example:
if ok := ValidateBackingStoreType(bs); !ok {
RunValidation(bs) // this function name can be more acurate too...
return
}

WDYT?

Reply:
I think that if we change the name of `RunValidation` we could skip the `ok` var and do something like this:
```
if !bsv.GetValidationResult(bsv.ValidateBSEmptySecretName()) {
return
}
```

Reply:
hmm I find it confusing to call it getValidationResult when the other one actually gets the result
let's discuss it on a short call :)